src/pkg/packager/common.go

```diff
@@ -440,6 +443,75 @@ func (p *Packager) validatePackageChecksums() error {
 	return nil
 }
 
+// validatePackageArchitecture validates that the package architecture matches the target cluster architecture.
+func (p *Packager) validatePackageArchitecture() error {
+	var (
+		applianceMode bool
+		clusterArch   string
+		err           error
+		k8sTarget     bool
+	)
+
+	// Iterate over the package components.
+	components := p.getValidComponents()
+	for _, component := range components {
+		// Determine whether we are working with an appliance mode init package.
+		if component.Name == k8s.DistroIsK3s && p.cfg.Pkg.Kind == "ZarfInitConfig" {
+			applianceMode = true
+		}
+
+		// Determine whether we are deploying k8s resources.
+		if component.Images != nil {
+			k8sTarget = true
+		}
+	}
+
+	// If we're working with an init package deploying k3s(appliance mode),
+	// set the clusterArch to the machine we're running on.
+	if applianceMode {
+		clusterArch = runtime.GOARCH
+
+		if p.arch != clusterArch {
+			return fmt.Errorf(lang.CmdPackageDeployValidateArchitectureErr, p.arch, clusterArch)
+		}
+
+		return nil
+	}
+
+	/*
+		If we're already connected to a cluster, query the cluster for the architecture.
+
+		If we're not already connected to a cluster and we're deploying k8s resources,
+		attempt to establish a new k8s client connection and query the cluster for the architecture.
+	*/
+	if p.cluster != nil {
+		clusterArch, err = p.cluster.Kube.GetArchitecture()
+		if err != nil {
+			return err
+		}
+
+		if p.arch != clusterArch {
+			return fmt.Errorf(lang.CmdPackageDeployValidateArchitectureErr, p.arch, clusterArch)
+		}
```

Comment:
This is dead code - `p.cluster` is always `nil` here as it won't get set until lines 153 or 224 of `deploy.go`.

Now that we have it (it wasn't there when this PR started) I would look at the following in `interactive.go`:

https://github.com/defenseunicorns/zarf/blob/5666ae237cb088628d40af0b597e8fe4f38933af/src/pkg/packager/interactive.go#L40

We don't use p.cluster there since we are only trying to see if we have access and may not at that time but that is a way to get access to a cluster early for pre-checks like this.

Reply:
Changes made